plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1846,6 +1868,61 @@ private void prepareAlterPartition(HdfsContext hdfsContext, HiveIdentity identit
                     new PartitionWithStatistics(oldPartition.get(), partitionName, oldPartitionStatistics)));
         }
 
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, PartitionAndMore partitionAndMore)
+        {
+            if (!partitionAndMore.isRetriesEnabled()) {
+                return;
+            }
+            if (!partitionAndMore.hasFileNames()) {
+                return;
+            }
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, partitionAndMore.getCurrentLocation(), ImmutableSet.copyOf(partitionAndMore.getFileNames()));
+        }
+
+        private void handleFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, TableAndMore tableAndMore)
+        {
+            if (!tableAndMore.isRetriesEnabled()) {
+                return;
+            }
+            if (tableAndMore.getCurrentLocation().isEmpty()) {
+                return;
+            }
+            if (tableAndMore.getFileNames().isEmpty()) {
+                return;
+            }
+
+            Path tableLocation = tableAndMore.getCurrentLocation().orElseThrow(() -> new IllegalArgumentException("table location not set"));
+            List<String> files = tableAndMore.getFileNames().orElseThrow(() -> new IllegalStateException("file list not not set"));
+            deleteFilesForFailedRetryAttempts(hdfsContext, queryId, tableLocation, ImmutableSet.copyOf(files));
+        }
+
+        private void deleteFilesForFailedRetryAttempts(HdfsContext hdfsContext, String queryId, Path path, Set<String> files)
+        {
+            ImmutableList.Builder<String> deletedFilesBuilder = ImmutableList.builder();
+            try {
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(hdfsContext, path);
+                if (!fileSystem.exists(path)) {
```

Comment:
Shouldn't it be a failure? (if the output directory doesn't exist)?

Reply:
I don't think so. We are not sure if there are any remaining files to be deleted. And in S3 empty directory would disappear if there is no marker file in it.

Reply:
Oh never mind. I wrote before I thought. It is target dir. Then yeah - I think it should be present always. Let me change that.